### PR TITLE
Fix missing chat messages

### LINKED ISSUE #12062: 2.3-beta4: Public chat doesn't show old messages after being locked
**Describe the bug**
If public chat is being locked, old messages aren't shown anymore for locked attendees

**To Reproduce**
1. Open session, log in as a mod and two attendees
2. Let attendee private chat with other attendee (for sure there are also other scenarios in which this happens, but this one is reproducable easily)
3. Lock public chat as mod
4. Click "public chat" as attendee (don't close private chat separately before)
5. All messages from before aren't shown for the attendee switching from private to public chat, sometimes it even cuts off default welcome message (see screenshot). This cut-off is not completely easy to reproducable, though.

**Expected behavior**
Old messages should be shown in the public chat as only **writing** messages should be locked.

**Actual behavior**
Public chat has completely been cleared for client, only shows messages sent after being locked.

**Screenshots**
grafik

**BBB version (optional):**
2.3-beta4

**Desktop (please complete the following information):**
Not browser related I feel like. I used current FF on Win10 for mod, current Chrome on Win10 for Clients

**Additional context**
Sometimes the red warning message that chat is locked is also not shown. I couldn't reproduce it reliably though.

### PR DESCRIPTION
**Pull request by ramonlsouza**
### What does this PR do?

- Fixes a bug where the public chat messages could disappear after the creation of a new private chat
- Updates messages height when switching between public and private chat 

### Closes Issue(s)
Closes #12062

### CODE REVIEW

**Comment by schrd** [comment]
I could trigger a crash with this PR. I applied this PR on top of (d96424dc82e8dedbe1952ce24ec0f9e22371416c).

I joined 1 Moderator, 1 Viewer. Bot chatting a lot in public chat. Moderator opens private chat with viewer → crash.

~~~
Error: Requested index 0 is outside of range 0..0
    e https://BBB-HOST/html5client/ce98534983f0b8ee0863406186063c087121f996.js?meteor_js_resource=true:190
    e https://BBB-HOST/html5client/ce98534983f0b8ee0863406186063c087121f996.js?meteor_js_resource=true:190
    e https://BBB-HOST/html5client/ce98534983f0b8ee0863406186063c087121f996.js?meteor_js_resource=true:190
    e https://BBB-HOST/html5client/ce98534983f0b8ee0863406186063c087121f996.js?meteor_js_resource=true:190
    e https://BBB-HOST/html5client/ce98534983f0b8ee0863406186063c087121f996.js?meteor_js_resource=true:190
    e https://BBB-HOST/html5client/ce98534983f0b8ee0863406186063c087121f996.js?meteor_js_resource=true:190
    e https://BBB-HOST/html5client/ce98534983f0b8ee0863406186063c087121f996.js?meteor_js_resource=true:190
    os https://BBB-HOST/html5client/ce98534983f0b8ee0863406186063c087121f996.js?meteor_js_resource=true:69
    ns https://BBB-HOST/html5client/ce98534983f0b8ee0863406186063c087121f996.js?meteor_js_resource=true:69
    Sl https://BBB-HOST/html5client/ce98534983f0b8ee0863406186063c087121f996.js?meteor_js_resource=true:69
    Hl https://BBB-HOST/html5client/ce98534983f0b8ee0863406186063c087121f996.js?meteor_js_resource=true:69
    Bl https://BBB-HOST/html5client/ce98534983f0b8ee0863406186063c087121f996.js?meteor_js_resource=true:69
    Al https://BBB-HOST/html5client/ce98534983f0b8ee0863406186063c087121f996.js?meteor_js_resource=true:69
    di https://BBB-HOST/html5client/ce98534983f0b8ee0863406186063c087121f996.js?meteor_js_resource=true:69
    unstable_runWithPriority https://BBB-HOST/html5client/ce98534983f0b8ee0863406186063c087121f996.js?meteor_js_resource=true:87
    ui https://BBB-HOST/html5client/ce98534983f0b8ee0863406186063c087121f996.js?meteor_js_resource=true:69
    di https://BBB-HOST/html5client/ce98534983f0b8ee0863406186063c087121f996.js?meteor_js_resource=true:69
    pi https://BBB-HOST/html5client/ce98534983f0b8ee0863406186063c087121f996.js?meteor_js_resource=true:69
    Cl https://BBB-HOST/html5client/ce98534983f0b8ee0863406186063c087121f996.js?meteor_js_resource=true:69
    enqueueForceUpdate https://BBB-HOST/html5client/ce98534983f0b8ee0863406186063c087121f996.js?meteor_js_resource=true:69
    forceUpdate https://BBB-HOST/html5client/ce98534983f0b8ee0863406186063c087121f996.js?meteor_js_resource=true:45
    computation https://BBB-HOST/html5client/ce98534983f0b8ee0863406186063c087121f996.js?meteor_js_resource=true:295
    _compute https://BBB-HOST/html5client/ce98534983f0b8ee0863406186063c087121f996.js?meteor_js_resource=true:257
    _recompute https://BBB-HOST/html5client/ce98534983f0b8ee0863406186063c087121f996.js?meteor_js_resource=true:257
    _runFlush https://BBB-HOST/html5client/ce98534983f0b8ee0863406186063c087121f996.js?meteor_js_resource=true:257
    a https://BBB-HOST/html5client/ce98534983f0b8ee0863406186063c087121f996.js?meteor_js_resource=true:1
ce98534983f0b8ee0863406186063c087121f996.js:69:83125
Exception from Tracker recompute function: ce98534983f0b8ee0863406186063c087121f996.js:1:6701
Error: Requested index 0 is outside of range 0..0 ce98534983f0b8ee0863406186063c087121f996.js:1:6701
e@https://BBB-HOST/html5client/ce98534983f0b8ee0863406186063c087121f996.js?meteor_js_resource=true:190:167684
e@https://BBB-HOST/html5client/ce98534983f0b8ee0863406186063c087121f996.js?meteor_js_resource=true:190:169543
e@https://BBB-HOST/html5client/ce98534983f0b8ee0863406186063c087121f996.js?meteor_js_resource=true:190:170085
e@https://BBB-HOST/html5client/ce98534983f0b8ee0863406186063c087121f996.js?meteor_js_resource=true:190:169169
e@https://BBB-HOST/html5client/ce98534983f0b8ee0863406186063c087121f996.js?meteor_js_resource=true:190:165440
e@https://BBB-HOST/html5client/ce98534983f0b8ee0863406186063c087121f996.js?meteor_js_resource=true:190:152544
e@https://BBB-HOST/html5client/ce98534983f0b8ee0863406186063c087121f996.js?meteor_js_resource=true:190:150771
os@https://BBB-HOST/html5client/ce98534983f0b8ee0863406186063c087121f996.js?meteor_js_resource=true:69:70578
ns@https://BBB-HOST/html5client/ce98534983f0b8ee0863406186063c087121f996.js?meteor_js_resource=true:69:70373
Sl@https://BBB-HOST/html5client/ce98534983f0b8ee0863406186063c087121f996.js?meteor_js_resource=true:69:105907
Hl@https://BBB-HOST/html5client/ce98534983f0b8ee0863406186063c087121f996.js?meteor_js_resource=true:69:97659
Bl@https://BBB-HOST/html5client/ce98534983f0b8ee0863406186063c087121f996.js?meteor_js_resource=true:69:97582
Al@https://BBB-HOST/html5client/ce98534983f0b8ee0863406186063c087121f996.js?meteor_js_resource=true:69:94500
di/<@https://BBB-HOST/html5client/ce98534983f0b8ee0863406186063c087121f996.js?meteor_js_resource=true:69:46016
e/r.unstable_runWithPriority@https://BBB-HOST/html5client/ce98534983f0b8ee0863406186063c087121f996.js?meteor_js_resource=true:87:3854
ui@https://BBB-HOST/html5client/ce98534983f0b8ee0863406186063c087121f996.js?meteor_js_resource=true:69:45725
di@https://BBB-HOST/html5client/ce98534983f0b8ee0863406186063c087121f996.js?meteor_js_resource=true:69:45963
pi@https://BBB-HOST/html5client/ce98534983f0b8ee0863406186063c087121f996.js?meteor_js_resource=true:69:45896
Cl@https://BBB-HOST/html5client/ce98534983f0b8ee0863406186063c087121f996.js?meteor_js_resource=true:69:91266
enqueueForceUpdate@https://BBB-HOST/html5client/ce98534983f0b8ee0863406186063c087121f996.js?meteor_js_resource=true:69:49999
e/_.prototype.forceUpdate@https://BBB-HOST/html5client/ce98534983f0b8ee0863406186063c087121f996.js?meteor_js_resource=true:45:1406
calculateData/this.computation</<@https://BBB-HOST/html5client/ce98534983f0b8ee0863406186063c087121f996.js?meteor_js_resource=true:295:1731
_compute@https://BBB-HOST/html5client/ce98534983f0b8ee0863406186063c087121f996.js?meteor_js_resource=true:257:2279
_recompute@https://BBB-HOST/html5client/ce98534983f0b8ee0863406186063c087121f996.js?meteor_js_resource=true:257:2434
n/o._runFlush@https://BBB-HOST/html5client/ce98534983f0b8ee0863406186063c087121f996.js?meteor_js_resource=true:257:3443
a@https://BBB-HOST/html5client/ce98534983f0b8ee0863406186063c087121f996.js?meteor_js_resource=true:1:3087
~~~

**Comment by schrd** [comment]
I extracted the data from mongodb if that helps:
[pr-12102-mongodb.txt](https://github.com/bigbluebutton/bigbluebutton/files/6351583/pr-12102-mongodb.txt)

**Comment by ramonlsouza** [comment]
> I could trigger a crash with this PR. I applied this PR on top of ([d96424d](https://github.com/bigbluebutton/bigbluebutton/commit/d96424dc82e8dedbe1952ce24ec0f9e22371416c)).
> 
> I joined 1 Moderator, 1 Viewer. Bot chatting a lot in public chat. Moderator opens private chat with viewer → crash.


fixed in 8f3a3a8

File: bigbluebutton-html5/imports/ui/components/chat/time-window-list/component.jsx [hunk 1]
```diff
@@ -109,6 +109,11 @@ class TimeWindowList extends PureComponent {
     const prevLastTimeWindow = prevTimeWindowsValues[prevTimeWindowsLength - 1];
     const lastTimeWindow = timeWindowsValues[prevTimeWindowsLength - 1];
 
+    if (prevLastTimeWindow?.chatId !== lastTimeWindow?.chatId) {
+      this.systemMessageIndexes = [];
+      this.listRef.forceUpdateGrid();
```
Comment (Tainan404) [inline bigbluebutton-html5/imports/ui/components/chat/time-window-list/component.jsx:114]: in the same componenDidUpdate we've the same logic of force the virtualized list update when changes the [chatId](https://github.com/ramonlsouza/bigbluebutton/blob/develop/bigbluebutton-html5/imports/ui/components/chat/time-window-list/component.jsx#L1316), why did you duplicate it here?

File: bigbluebutton-html5/imports/ui/components/components-data/chat-context/adapter.jsx [hunk 2]
```diff
@@ -62,10 +62,14 @@ const Adapter = () => {
...
 
   useEffect(() => {
-    window.addEventListener(EVENT_NAME, () => {
-      dispatch({
-        type: ACTIONS.CLEAR_STREAM_MESSAGES,
-      });
+    window.addEventListener(EVENT_NAME, (e) => {
+      const { role } = e.detail;
+
+      if (role !== usersData[Auth.userID].role) {
```
Comment (Tainan404) [inline bigbluebutton-html5/imports/ui/components/components-data/chat-context/adapter.jsx:68]: What's the role impact in messages cleaning?

File: bigbluebutton-html5/imports/ui/components/chat/time-window-list/component.jsx (hunk 1 above)
Comment (ramonlsouza) [reply bigbluebutton-html5/imports/ui/components/chat/time-window-list/component.jsx]: you're right, no need to use `forceUpdateGrid` twice, but we need to recompute heights when switching between public and private chat... fixed in c9540df

File: bigbluebutton-html5/imports/ui/components/components-data/chat-context/adapter.jsx (hunk 2 above)
Comment (ramonlsouza) [reply bigbluebutton-html5/imports/ui/components/components-data/chat-context/adapter.jsx:68]: #12012 added the messages cleaning, but we only needed to clean messages if the user role changes (#11842)

Comment (Tainan404) [reply bigbluebutton-html5/imports/ui/components/components-data/chat-context/adapter.jsx:68]: Taking a little more attention, it looks like there may be a race condition, as you are using to compare the role from user collection and role from user context. I think the information may take a little longer to update to the context, but there is no 100% guarantee that this logic will work every time.

Comment (ramonlsouza) [reply bigbluebutton-html5/imports/ui/components/components-data/chat-context/adapter.jsx]: moved role check to `chat-context/adapter` in 1a032e4

**Comment by schrd** [comment]
I applied this on top of develop (1416a0d255e4ebc43eb1f18594d85e92fcbb508e). Messages don't disappear anymore, but there is still an issue with the new messages counter.

I could reproduce using the following steps:
Join as 3 users:

* A as moderator
* B as viewer
* C as viewer

Now send messages:

* B sends a message to public chat. User B keeps public chat open.
* C sends a private message to B
* A sends private message to B

read messages in the following order

* B reads messages from A and C
* All messages are read.
* A promotes B to moderator

Message counters of user B are broken and show messages as unread: usually the messages of A and C, sometimes also in public chat.

**Comment by ramonlsouza** [comment]
> I applied this on top of develop ([1416a0d](https://github.com/bigbluebutton/bigbluebutton/commit/1416a0d255e4ebc43eb1f18594d85e92fcbb508e)). Messages don't disappear anymore, but there is still an issue with the new messages counter.
> 
> I could reproduce using the following steps:
> Join as 3 users:
> 
> * A as moderator
> * B as viewer
> * C as viewer
> 
> Now send messages:
> 
> * B sends a message to public chat. User B keeps public chat open.
> * C sends a private message to B
> * A sends private message to B
> 
> read messages in the following order
> 
> * B reads messages from A and C
> * All messages are read.
> * A promotes B to moderator
> 
> Message counters of user B are broken and show messages as unread: usually the messages of A and C, sometimes also in public chat.

fixed in adae0f2

**Comment by antobinary** [comment]
@Tainan404 anything else missing?

**Comment by Tainan404** [comment]
@antobinary, nothing to add

**Comment by antobinary** [comment]
@daminebenq could you please test one more time

**Comment by daminebenq** [comment]
Looks good. Couldn't reproduce anything wrong with this PR.

**Pull request merged**